[quant][fx] Add support for matching multiple arguments in patterns

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89986

Summary:
This PR adds support for matching patterns that has multiple arguments, it's needed for quantization in PyTorch 2.0 early prototype

Before this PR, we only support patterns like:
```
x -> conv -> bn -> relu
(relu, (bn, conv))
```
where each operator has a single node, the code breaks when we want to match a pattern that has an op that has multiple arguments, such as:
```
                           shape \
        transpose -> reshape -> output ->
```
where `reshape` has two arguments

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_match_pattern_with_multiple_args

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

jerryzh168 committed:
[quant][fx] Add support for matching multiple arguments in patterns

Summary:
This PR adds support for matching patterns that has multiple arguments, it's needed for quantization in PyTorch 2.0 early prototype

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_match_pattern_with_multiple_args

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for matching multiple arguments in patterns"


Summary:
This PR adds support for matching patterns that has multiple arguments, it's needed for quantization in PyTorch 2.0 early prototype

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_match_pattern_with_multiple_args

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -241,10 +241,13 @@ def _is_pattern_dtype_config_and_qconfig_supported_by_backend(
     # TODO: this only works for one input and one output patterns, need to generalize to multiple
     # inputs/output
-    root_node = _default_root_node_getter(matched_node_pattern)
+    # root_node = _default_root_node_getter(matched_node_pattern)
```
delete?

vkuzo commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5479,6 +5479,54 @@ def forward(self, x):
             self.checkGraphModuleNodes(m_ref, expected_node_occurrence=node_occurrence_ref)
 
+    def test_match_pattern_with_multiple_args(self):
+        """ Test that we can match a pattern that has multiple arguments
```
what does "multiple arguments" mean?

jerryzh168 replied:
this means `reshape` has multiple arguments, also added in the docstring

vkuzo commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5479,6 +5479,54 @@ def forward(self, x):
+                return transpose
+
+            backend_pattern_configs.append(
+                BackendPatternConfig((torch.reshape, torch.transpose, MatchAllNode))
```
this looks like a pattern of a -> b -> c, could we explain what specifically is different about this compare to what's already supported?

jerryzh168 replied:
this means `a(b(input1), c(input2))` actually, sure, I'll add more explanations

jerryzh168 committed:
Update on "[quant][fx] Add support for matching multiple arguments in patterns"


Summary:
This PR adds support for matching patterns that has multiple arguments, it's needed for quantization in PyTorch 2.0 early prototype

Before this PR, we only support patterns like:
```
x -> conv -> bn -> relu
(relu, (bn, conv))
```
where each operator has a single node, the code breaks when we want to match a pattern that has an op that has multiple arguments, such as:
```
                           shape \
        transpose -> reshape -> output ->
```
where `reshape` has two arguments

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_match_pattern_with_multiple_args

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
